deepspeed/runtime/activation_checkpointing/checkpointing.py

```diff
@@ -365,11 +365,14 @@ def partition_activations(args, cpu_checkpoint, contiguous_checkpoint):
     global contiguous_data_buffers, data_offsets
 
     inputs = []
+    num_non_fp_tensors = 0
     for i, item in enumerate(args):
         if not is_activation_to_checkpoint(item):
             inputs.append(item)
+            num_non_fp_tensors += 1
             continue
 
+        i -= num_non_fp_tensors

```

Comment:
Can we avoid modifying the index variable `i`? This pattern makes the code brittle. Will the following pattern work?

```python
i = 0
for item in args:
  # loop body
  i+1
```

Reply:
At the bottom, `i` is used, and that's where the problem arises.

Reply:
I know it's a bad idea to change `i`, but it's the best way to maintain the overall code structure.

Reply:
`i` is used very often. Are you sure you want to change all these `i`?

```python
if contiguous_checkpoint:
            if i >= len(contiguous_data_buffers):
                tensor_list = [
                    torch.tensor(()).new_empty([partition_size],
                                               dtype=partition.dtype,
                                               device=buffer_device)
                    for i in range(num_layers)
                ]
                contiguous_data_buffers.append(tensor_list)
                data_offsets.append(0)
            elif contiguous_data_buffers[i] is None:
                tensor_list = [
                    torch.tensor(()).new_empty([partition_size],
                                               dtype=partition.dtype,
                                               device=buffer_device)
                    for i in range(num_layers)
                ]
                contiguous_data_buffers[i] = tensor_list
                data_offsets[i] = 0

            # Because the 'new_empty' returns uninitialized pages,
            # the pages need to be populated during the cudaMemcpy time
            # which increases the data copy time. To avoid this, we
            # pre-populate these pages by simply writing 0 ahead of
            # the actual cudaMemcpy operation time. Due to the
            # previously launched GPU kernels, there is a small
            # window of time here for CPUs to populate pages asynchronously.
            contiguous_data_buffers[i][data_offsets[i]].data[range(
                0,
                contiguous_data_buffers[i][data_offsets[i]].data.shape[0],
                int(mmap.PAGESIZE /
                    contiguous_data_buffers[i][data_offsets[i]].data.element_size())
            )] = 0

            contiguous_partition = contiguous_data_buffers[i][
                data_offsets[i]].data.copy_(partition.data)
            data_offsets[i] = data_offsets[i] + 1
            inputs.append(contiguous_partition)
        else:
            partition = partition.cpu() if CPU_CHECKPOINT else partition
            inputs.append(partition)
```

Reply:
Also, the `i` used in the outer loop is used again in the inner loop. It seems good to change the `i` of the inner loop to `j`.

```python
    for i, item in enumerate(args):
        if not is_activation_to_checkpoint(item):
            inputs.append(item)
            num_non_fp_tensors += 1
            continue

        i -= num_non_fp_tensors
        partition_size = get_partition_size(item)
        partition = item.detach().contiguous().view(-1).narrow(
            0,
            get_partition_start(item),
            partition_size).clone()

        buffer_device = torch.device('cpu') if cpu_checkpoint else partition.device

        if contiguous_checkpoint:
            if i >= len(contiguous_data_buffers):
                tensor_list = [
                    torch.tensor(()).new_empty([partition_size],
                                               dtype=partition.dtype,
                                               device=buffer_device)
                    for i in range(num_layers)
                ]
                contiguous_data_buffers.append(tensor_list)
                data_offsets.append(0)
```

Reply:
I agree that `i` is used often and it would be too painful to change it. This is why I proposed initializing `i` outside the loop and incrementing appropriately in the loop. That way the rest of the code does not have to change. Does that make sense?

Reply:
yes i think it's good By the way, as you suggested, I split the PR, so I think it would be good to discuss it #1376.